python/ray/tune/suggest/hyperopt.py

```diff
@@ -0,0 +1,117 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+import copy
+try:
+    import hyperopt as hpo
+except Exception as e:
+    hpo = None
+
+from ray.tune.error import TuneError
+from ray.tune.suggest import ExistingVariants
+
+
+class HyperOptSearch(ExistingVariants):
+    """A wrapper around HyperOpt to provide trial suggestions.
+
+    Requires HyperOpt to be installed from source.
+    Uses the Tree-structured Parzen Estimators algorithm, although can be
+    trivially extended to support any algorithm HyperOpt uses. Externally
+    added trials will not be tracked by HyperOpt.
+
+    Parameters:
+        space (dict): HyperOpt configuration. Parameters will be sampled
+            from this configuration and will be used to override
+            parameters generated in the variant generation process.
+        max_concurrent (int): Number of maximum concurrent trials. Defaults
+            to 10.
+        reward_attr (str): The TrainingResult objective value attribute.
+            This refers to an increasing value, which is internally negated
+            when interacting with HyperOpt so that HyperOpt can "maximize"
+            this value.
+        experiments (Experiment | list | dict): Experiments to run. Will be
+            used by ExistingVariants parent class to initialize Trials.
+    """
+
+    def __init__(self,
+                 space,
+                 max_concurrent=10,
+                 reward_attr="episode_reward_mean",
+                 experiments=None,
+                 **kwargs):
+        assert hpo is not None, "HyperOpt must be installed!"
+        assert type(max_concurrent) is int and max_concurrent > 0
+        self._max_concurrent = max_concurrent
+        self._reward_attr = reward_attr
+        self.algo = hpo.tpe.suggest
+        self.domain = hpo.Domain(lambda spc: spc, space)
+        self._hpopt_trials = hpo.Trials()
+        self._live_trials = {}
+        self.rstate = np.random.RandomState()
+
+        super(HyperOptSearch, self).__init__(experiments=experiments, **kwargs)
+
+    def try_suggest(self):
```

Comment:
Instead of returning a trial id, pass in a trial id. That way, the return is just the config or None, no need for PASS?

Reply:
I' thinking what if the algorithm want to return a bunch of trials. Currently, it would be very clumsy to do this (i.e. store a bunch of trials inside and return them one by one)

Reply:
I think that's a good idea; however, I think that should not go in here (i.e., HyperBand). 

I think `next_trials` should be modified to return a list of trials. After all, this method isn't actually exposed to the rest of Tune.

Reply:
Also, the point in reference from @old-bear is for genetic algorithms (and also using their own searchspace).